Cypress Migration Guide (#8942)
department-of-veterans-affairs/va.gov-team: teams/vsp/teams/tools/frontend/cypress-migration-guide.md

Nightwatch:
```javascript
client.selectDropdown('root_veteranAddress_country', data.veteranAddress.country)
```
Cypress:
```javascript
cy.get('#root_veteranAddress_state').select(testData.veteranAddress.state)
```

#### Entering Data
Nightwatch:
```javascript
client.fill('input[name="root_firstName"]', data.veteranFullName.first)
```
Cypress:
```javascript
cy.get('#root_firstName').type(testData.veteranFullName.first)
```

Please visit [here](https://docs.cypress.io/guides/core-concepts/interacting-with-elements.html#Actionability) for more information about interacting with elements. For additional ways to interact with the DOM we have also included the [cypress testing library](https://testing-library.com/docs/cypress-testing-library/intro) in the repo.

### Custom Commands <a name="custom-commands"></a>

Some custom nightwatch commands located in `src/platform/testing/e2e/nightwatch-commands` were converted to cypress commands, and can be found in `src/platform/testing/e2e-cypress/support`. Many of them did not need to be converted because cypress supports their functionality natively. These custom commands can be used directly on `cy`. For example:

```javascript

> cy.fillDate('root_dob', testData.veteranDateOfBirth);

Reviewer: Just realized this is also used as an example for the Accessibility section later... So not sure what a good example would be, since we're not using `fillDate`. Maybe it's fine to repeat.

> cy.fillDate('root_dob', testData.veteranDateOfBirth);

Author: Good catch. I think it's fine to repeat.

To add axe checks to your tests use the custom `axeCheck()` command based off the `checkA11y` [command](https://github.com/avanslaars/cypress-axe#cychecka11y).

#### Nightwatch:
```javascript
client.axeCheck('.main')
```

#### Cypress:
```javascript
cy.axeCheck();
```

### Assertions <a name="assertions"></a>

Cypress uses `chai` to [handle assertions](https://docs.cypress.io/guides/references/assertions.html#BDD-Assertions). Currently in the `nightwatch-assertions` folder, we have 4 custom assertions, which can all be replaced:

#### hasFocusableCount / hasTabbableCount
Checks the count of focusable and tabbable elements.

These can be converted by making [custom chai assertions](https://docs.cypress.io/guides/references/assertions.html#Adding-New-Assertions).

#### isActiveElement
Checks if the given element is focused on the page.

```javascript
// Bdd assertion
cy.get('input').should('have.focus')
```
#### isDisabledElement 
Checks if the given element is disabled on the page.

```javascript
// Bdd assertion
cy.get('button').should('be.disabled')
```

## Running Tests <a name="running-tests"></a>

> As of now, cypress supports Chrome, Edge, Firefox(beta) and a few [others](https://docs.cypress.io/guides/guides/launching-browsers.html#Browsers). You can run tests in headless mode or via the test runner.

Reviewer: ```suggestion
Cypress supports Chrome, Edge, Firefox, and a few [others](https://docs.cypress.io/guides/guides/launching-browsers.html#Browsers). You can run tests in headless mode or via the test runner.
```

### Headless mode <a name="headless-mode"></a>

> To run tests headless, run `yarn test:cypress:run`

Reviewer: ```suggestion
To run tests headless, run `yarn cypress run --headless`
```

### Test Runner <a name="test-runner"></a>

> To run tests using the cypress [test runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview), run `yarn test:cypress:open`.

Reviewer: ```suggestion
To run tests using the Cypress [test runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview), run `yarn cypress open`.
```

> With the Test Runner, you are able to use cypress's `Open Selector Playground'. This allows you to click on elements in the DOM and copy the selector for that item for your test. This means there's usually no need to inspect an element manually.

Reviewer: ```suggestion
With the test runner, you are able to use Cypress's "Open Selector Playground". This allows you to click on elements in the DOM and copy that element's selector to use in your test. This means there's usually no need to manually inspect an element and copy the selector.
```

The test runner provides the ability to pause tests, and [time travel](https://docs.cypress.io/guides/getting-started/writing-your-first-test.html#Time-travel), which allows you to see snapshots of your tests.

You can add any of the [run options](https://docs.cypress.io/guides/guides/command-line.html#Commands) to the commands above.

## Things to Note <a name="things-to-note"></a>

### Automatic Waiting

> Cypress automatically waits for commands to execute before moving on to the next. This eliminates needing to use the timeout consts in `platform/testing/e2e/timeouts.js`

Reviewer: No example off the top of my head, but I think it might be worth pointing out that Cypress queues up its commands instead of runniing them synchronously. Clearing that up for people might help them avoid a painful ordeal with trying to combine Cypress commands with synchronous code.

Reviewer: [This might work.](https://docs.cypress.io/guides/references/best-practices.html#Assigning-Return-Values)

- `plugins` contains custom tasks, which allow you to [tap into the node environment](https://docs.cypress.io/guides/tooling/plugins-guide.html). This is valuable because all cypress test code is executed in the browser, so plugins allow us to execute code in the node process running outside of the browser.
- `support` holds [custom cypress commands](https://docs.cypress.io/api/cypress-api/custom-commands.html#Syntax), specifically in the `commands.js` file. These commands can be run in cypress tests, similar to the built in commands. This is similar to [Nightwatch's custom commands](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/common-tasks/new-end-to-end-test/#custom-nightwatch-commands). This is also where cypress plugins are imported, as shown in `index.js`.

### Test Dependencies
You do not need to import any modules for helpers and timeouts, etc as before with nightwatch. You can start writing your tests without anything needed to be imported.

## Writing Tests <a name="writing-tests"></a>
### Test Structure <a name="test-structure"></a>

> The test structure of Cypress should feel familiar. Cypress uses mocha's bdd syntax which is what we use for our unit tests. Note that because `it()` blocks are their own tests, each `it()` block should be independent of others in the same testing suite. Everything executed in the browser needs to be inside of an `it()` block. Visit [here](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Test-Structure) for more context.

Reviewer: Each spec file spawns a new browser instance. Each `it` is isolated in setup and such but can all be inspected within an individual spec in one browser instance.

```
Cypress:
```javascript
cy.get('#root_veteranAddress_state').select(testData.veteranAddress.state)
```

#### Entering Data
Nightwatch:
```javascript
client.fill('input[name="root_firstName"]', data.veteranFullName.first)
```
Cypress:
```javascript
cy.get('#root_firstName').type(testData.veteranFullName.first)
```

Please visit [here](https://docs.cypress.io/guides/core-concepts/interacting-with-elements.html#Actionability) for more information about interacting with elements. For additional ways to interact with the DOM we have also included the [cypress testing library](https://testing-library.com/docs/cypress-testing-library/intro) in the repo.

### Custom Commands <a name="custom-commands"></a>

Some custom nightwatch commands located in `src/platform/testing/e2e/nightwatch-commands` were converted to cypress commands, and can be found in `src/platform/testing/e2e-cypress/support`. Many of them did not need to be converted because cypress supports their functionality natively. These custom commands can be used directly on `cy`. For example:

```javascript
cy.fillDate('root_dob', testData.veteranDateOfBirth);
```

### Test Data <a name="test-data"></a>

> In Cypress, because everything in a test is executed inside of the browser, [fixtures](https://docs.cypress.io/api/commands/fixture.html#Syntax) are used to get access to data in tests.

Reviewer: Not natively, but I've come up with a way to sync arbitrary paths and files as fixtures.

Fixtures should be stored in the app specific folder inside of `src/platform/testing/e2e-cypress/fixtures`. If the app specific folder does not exist feel free to create it.

For example: `src/platform/testing/e2e-cypress/fixtures/hca` contains all the fixtures for the `hca` application. You can copy all the schema files used in current tests to the proper fixtures folder.

You can inject the fixture into your test by doing:
```javascript
cy.fixture('hca/maximal-test.json').as('testData');
```
Cypress already knows to look in the `fixtures` folder, so you can use the application specific directory for the path to the fixture.

To get access to the fixture, you can use `.get()`:

```javascript
cy.get('@testData').then(testData => {
  cy.get('#root_firstName').type(testData.veteranFullName.first);
});
```

### Mocking Data <a name="mocking-data"></a>

> Cypress allows you to [stub responses](https://docs.cypress.io/guides/guides/network-requests.html#Stubbing), which can avoid the need to use the mock api helpers.

Reviewer: They reset on every test, so stubs and aliases have to be done in a `beforeEach`.

        status: '404',
      },
    ],
  },
}).as('getApplication');

// Stub request to submit application
cy.route('POST', '/v0/health_care_applications', {
  formSubmissionId: '123fake-submission-id-567',
  timestamp: '2016-05-16',
}).as('submitApplication');

// Workaround to intercept requests made with fetch API.
cy.on('window:before:load', window => {
  delete window.fetch;
});
```
Based on the above stubs, whenever the browser makes a `GET` request to `/v0/health_care_applications/enrollment_status*` or a `POST` request to `/v0/health_care_applications`, cypress will automatically stub the given response.

All of your mock api calls using `mockData()` can be replaced with `cy.route()`. Note, be sure to always start the server with `cy.server()` before stubbing requests.

### File Uploads <a name="file-uploads"></a>
File uploads can be done using the [cypress file upload](https://github.com/abramenal/cypress-file-upload/tree/v3.5.3) plugin. For instructions on usage please visit [here](https://github.com/abramenal/cypress-file-upload/tree/v3.5.3#usage).

**Note:** Version 4 is unstable to please follow instructions for v3.5.3.

### Accessibility <a name="accessibility"></a>

> Just like most e2e tools, cypress has its own [axe-core plugin](https://github.com/avanslaars/cypress-axe) to test accessibility.

Author: It should deliver the same coverage as nightwatch because they are both using `axe` at their core. It may be helpful for us to go over the global `axe-core` settings in cypress with Trevor to see if he has any recommendations.

Many of them did not need to be converted, either because Cypress supports their functionality natively or the functionality was [too simple to abstract into a command](https://docs.cypress.io/api/cypress-api/custom-commands.html#2-Don%E2%80%99t-overcomplicate-things).

These custom commands can be called from the `cy` object. For example:

```javascript
cy.axeCheck();
```

### Test Data (Fixtures) <a name="test-data"></a>

In Cypress, because everything in a test is executed inside of the browser, [fixtures](https://docs.cypress.io/api/commands/fixture.html#Syntax) are used to get access to data in tests.

By default, Cypress doesn't support fixtures in separate directories, so we have a custom command for accessing fixtures stored in your app's directory.

For example, `src/applications/hca/tests/schema` contains test data for the `hca` application. You can load the file as a fixture like so:

```javascript
cy.syncFixtures({
  'data': 'srce/applications/hca/tests/schema',
  'minimal-test.json': 'src/applications/hca/tests/schema/maximal-test.json', 
  'maximal-test': 'src/applications/hca/tests/schema/maximal-test.json',
});

> cy.fixture('maximal-test.json').as('testData');

Reviewer: ```suggestion
cy.fixture('maximal-test').as('testData');
```

```

To access the contents of the file, you can use a combination of `cy.get()` and `cy.then()`:

```javascript
cy.get('@testData').then(testData => {
  cy.get('#root_firstName').type(testData.veteranFullName.first);
});
```

### Mocking Data <a name="mocking-data"></a>

Cypress allows you to [stub responses](https://docs.cypress.io/guides/guides/network-requests.html#Stubbing), avoiding the need for mock API helpers.

```javascript
// Start a server to begin routing responses
cy.server();

// Stub request to get enrollment status
cy.route({
  method: 'GET',
  url: '/v0/health_care_applications/enrollment_status*',
  status: 404,
  response: {
    errors: [
      {
        title: 'Record not found',
        detail: 'The record identified by  could not be found',
        code: '404',
        status: '404',
      },
    ],
  },
}).as('getApplication');

// Stub request to submit application
cy.route('POST', '/v0/health_care_applications', {
  formSubmissionId: '123fake-submission-id-567',
  timestamp: '2016-05-16',
}).as('submitApplication');

> 
> // Workaround to intercept requests made with fetch API.
> cy.on('window:before:load', window => {
>   delete window.fetch;
> });

Reviewer: ```suggestion
```

Devs won't need to add this to tests, since it's being done within the support file.

Cypress supports Chrome, Edge, Firefox, and a few [others](https://docs.cypress.io/guides/guides/launching-browsers.html#Browsers). You can run tests in headless mode or via the test runner.

### Headless mode <a name="headless-mode"></a>
To run tests headless, run `yarn cypress run --headless`

### Test Runner <a name="test-runner"></a>
To run tests using the Cypress [test runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview), run `yarn cypress open`.

With the test runner, you are able to use Cypress's "Open Selector Playground". This allows you to click on elements in the DOM and copy that element's selector to use in your test. This means there's usually no need to manually inspect an element and copy the selector.

The test runner provides the ability to pause tests, and [time travel](https://docs.cypress.io/guides/getting-started/writing-your-first-test.html#Time-travel), which allows you to see snapshots of your tests.

You can add any of the [run options](https://docs.cypress.io/guides/guides/command-line.html#Commands) to the commands above.

## Things to Note <a name="things-to-note"></a>

### Automatic Waiting

> Cypress automatically waits for commands to execute before moving on to the next. This eliminates needing to use the timeout consts in `platform/testing/e2e/timeouts.js`. Cypress queues it's commands instead of running them synchronously, so doing something like [this](https://docs.cypress.io/guides/references/best-practices.html#Assigning-Return-Values) will not work.

Reviewer: ```suggestion
Cypress automatically waits for commands to execute before moving on to the next one. This eliminates the need to use the timeout constants in `platform/testing/e2e/timeouts.js`.

Cypress queues its commands instead of running them synchronously, so doing something like [this](https://docs.cypress.io/guides/references/best-practices.html#Assigning-Return-Values) will not work.
```
